tracer/src/Datadog.Trace/RemoteConfigurationManagement/ProductConfigChangedEventArgs.cs

```diff
@@ -7,24 +7,63 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Text;
+using Datadog.Trace.Logging;
+using Datadog.Trace.RemoteConfigurationManagement.Protocol;
 using Datadog.Trace.Vendors.Newtonsoft.Json;
 
 namespace Datadog.Trace.RemoteConfigurationManagement
 {
     internal class ProductConfigChangedEventArgs : EventArgs
     {
-        private readonly IEnumerable<byte[]> _configContents;
+        private readonly IEnumerable<NamedRawFile> _configContents;
 
-        public ProductConfigChangedEventArgs(IEnumerable<byte[]> configContents)
+        private Dictionary<string, ApplyDetails> applyStates = new();
+
+        public ProductConfigChangedEventArgs(IEnumerable<NamedRawFile> configContents)
         {
             _configContents = configContents;
         }
 
-        public IEnumerable<T> GetDeserializedConfigurations<T>()
+        public IEnumerable<NamedTypedFile<T>> GetDeserializedConfigurations<T>()
         {
             return
                 _configContents
-                   .Select(bytes => JsonConvert.DeserializeObject<T>(Encoding.UTF8.GetString(bytes)));
+                   .Select(c => new NamedTypedFile<T>(c.Name, JsonConvert.DeserializeObject<T>(Encoding.UTF8.GetString(c.RawFile))));
+        }
+
+        public void Acknowldge(string filename)
+        {
+            var applyDetails = GetOrCreateApplyDetails(filename);
+
+            // can only move from unack to ack
+            if (applyDetails.ApplyState == ApplyStates.UNACKNOWLEDGED)
+            {
+                applyDetails.ApplyState = ApplyStates.ACKNOWLEDGED;
+            }
```

Comment:
What if it was an error before? we leave it as errored, even though it's now acknowledged? That differs with the behaviour in `RemoteConfigurationCache` AFAICT?

Reply:
This bit of code is all called if the remote config client detects a change in the message. I believe it actually plays nicely with resetting RemoteConfigurationCache , each time the messages change. On each remote config message change we:

1. reset the cache so ApplyState passes to Unack
2. attempt to apply all the new state to all listeners
3. update ApplyState according to what happened in the listener

However, I do agree this creates a coupling between the cache and the way ApplyState is handled that is rather delicate, but not sure how this could be improved?